### ISSUE TITLE
Allow project members to create namespaces in a project using kubectl

### ISSUE THREAD

**Issue by janeczku** [rancher/rancher#21222]
**What kind of request is this (question/bug/enhancement/feature request):**
Enhancement

**Steps to reproduce (least amount of steps as possible):**

As project member, I want to be able to use a native K8s API client implementation (e.g. `kubectl`) to manage resources in my project. This includes the ability to create new namespaces, e.g. in order to deploy manifests that require specific namespace scopes.

Creating namespaces as a project member via `kubectl` results in missing permission errors when trying to manage any resource there-in:

```
$ kubectl create ns new-ns
$ kubectl get pods -n new-ns
Error from server (Forbidden): pods is forbidden: User "u-xxxxx" cannot list pods in the namespace new-ns
```

In order to deploy resources in a new namespace as a project member using a native K8s client, one must do one of the following prerequisited steps:

1. Login to the Rancher GUI and pre-create the namespace in the specific project.
or
2. Use the `rancher` CLI to pre-create the namespace in the specific project context:
    - `$ rancher context switch`
    - `$ rancher namespace create ns-foo`

None of the above allows for a fully native K8s workflow and there are cases when even these workarounds won't work, e.g., when the client is an automated system (e.g. CI server) that supports K8s' but not Rancher's API.

A third (undocumented) workaround is to add an annotation to the namespace resource referencing the project ID. This has the following drawbacks:
1. Users need to remember a cryptic ID string instead of a project name
2. Manifests referencing project IDs instead of names are not portable across clusters

**Other details that may be helpful:**

Proposed implementation (requires project names to be unique in the cluster):

Users may annotate namespace resources with the project name. E.g.:

```
$ kubectl create ns foo && kubectl annotate ns/foo cattle.io/projectName=foo
```
or:
```
cat <<EOF | kubectl create -f -
apiVersion: v1
kind: Namespace
metadata:
  name: cassandra
  annotations:
    cattle.io/projectName: "backend"
EOF
```

**Environment information**
- Rancher version: v2.2.4
- Installation option (single install/HA): HA

**Comment by dawidmalina** [comment]
I agree this would be very interesting and desired improvement. This was also suggested (advised) in ticket: https://github.com/rancher/rancher/issues/16493

**Comment by dawidmalina** [comment]
Would be good if someone from rancher team tell us if such change is possible and considered as useful? @vincent99 @loganhz can you help with my question?

In our case it would be super useful to initially preconfigure custom cluster before import to rancher.

**Comment by vincent99** [comment]
This is not a simple thing to do because the projects don't "live" in the cluster at all, they are part of the management server.

**Comment by dawidmalina** [comment]
Yes I understand however if I annotate namespace with proper project id it will automatically be moved to project (and I'm doing it in the imported cluster not in the management server).
```
kubectl annotate ns/aggregators field.cattle.io/projectId=local:p-ts68x
```
So this is exactly what I would like to have but to be able annotate with `projectName` instead of `projectId`.

And would it at least be possible to automatically bind unannotated namespaces to default project?

**Comment by dawidmalina** [comment]
I'm not the rancher developer but https://github.com/rancher/rancher/blob/master/pkg/controllers/user/nslabels/labels.go looks like a good candidate and just extend current logic with project id lookup based on the project name. But someone with code knowledge must validate this.

**Comment by l-rossetti** [comment]
This feature would enforce a full integration between kubernetes and rancher. Following..

**Comment by dawidmalina** [comment]
@superseb, @vincent99, @ibuildthecloud any comments from rancher team? Looks like such improvment is highly desired and this is what community is looking for. I know there is rancher cli and we can move namespace to project but doing it with pure kubernetes cli (annotation) is at least for us is much better. It is now possible with project id so i don't see any reasons (issues) to extend this to support project name.

**Comment by stefanlasiewski** [comment]
From what I understand, there is too much work involved to `kubectl` to work Rancher's RBAC & 'Project* idea for all use cases. This is one reason why `rancher kubectl` itself exists.

In a nutshell, `kubectl` itself can be used by Cluster Admins and annotations such as `cattle.io/projectName: "backend"` will work. But if your users are constrained to Projects, they must use `rancher kubectl` instead.

**Comment by tomikmar** [comment]
This workaround seems to work: 

    kubectl create -f namespace-prod.yml
 
```
apiVersion: v1
kind: Namespace
metadata:
  name: prod123
  labels:
    name: prod
  annotations:
    # workaround for: https://github.com/rancher/rancher/issues/21222
    field.cattle.io/projectId: "c-abcde:p-12345"
```

Rancher version: v2.3.6

**Comment by chrisfosterelli** [comment]
The `field.cattle.io/projectId` is handy but only seems to work for those with "owner" permission to the cluster. Users who are only Members get an exception as `kubectl` can't read the namespace to see if it exists before creation (I presume). 

Is there a way to make the workaround work for cluster members?

**Comment by ericpromislow** [comment]
Here's a script:

```
#!/bin/bash

set -eu

case $# in
  4) ;;
  *) echo "Usage: $(basename $0) CLUSTERNAME PROJECT_NAME NEW_NAMESPACE_NAME DOWNSTREAM_CONFIG"
     echo "Something like: $(basename $0) mytestCluster proj01 newNS $HOME/Downloads/myKubeConfig.yaml"
     exit 1
esac

set -x

CLUSTERNAME=$1
PROJECT_NAME=$2
NEW_NAMESPACE_NAME=$3
DOWNSTREAM_CONFIG=$4

CLUSTER_ID="$(kubectl get clusters.management.cattle.io -o json |
  jq -r '.items[] |
         select(.metadata.annotations["field.cattle.io/description"] == "'$CLUSTERNAME'") | .metadata.name ')"

if [ -z "${CLUSTER_ID:-}" ]; then
   echo "Can't find cluster-id for cluster '$CLUSTERNAME'"
   exit 2
fi


PROJECT_ID=$(kubectl get projects -n $CLUSTER_ID -o json |
    jq -r '.items[] | select(.spec.displayName == "'$PROJECT_NAME'") | .metadata.name')

if [ -z "${PROJECT_ID:-}" ]; then
   echo "Can't find project-id for project $PROJECT_NAME in cluster '$CLUSTERNAME' (ID '$CLUSTER_ID')"
   exit 2
fi

echo "About to create namespace $NEW_NAMESPACE_NAME in cluster/project ${CLUSTER_ID}:${PROJECT_ID}"

KUBECONFIG=$DOWNSTREAM_CONFIG kubectl create -f - <<EOF
apiVersion: v1
kind: Namespace
metadata:
  name: $NEW_NAMESPACE_NAME
  annotations:
    field.cattle.io/projectId: ${CLUSTER_ID}:${PROJECT_ID}
EOF
```

The `$DOWNSTREAM_CONFIG` is the path to the kubeconfig file downloaded when logged in as a standard user and from the downstream cluster. The script assumes the default `kubeconfig` context is pointing at the kubeneretes cluster that hosts the local rancher cluster.

**Comment by johanot** [comment]
I understand Rancher projects are an _enrichment_ of the Kubernetes API and hence it is difficult to overload `kubectl create ns ...` without replacing kubectl with a porcelain cli.

Hence; I would actually be reasonably happy, if I could just limit my project members to create namespaces _within_ their own projects only. I believe it would then be fair to tell them, that namespace creation _has_ to happen through Rancher (either CLI or UI) and can't be done with kubectl.

However, currently, project members can use kubectl to create orphan namespaces, but they can't see them nor delete them once they are created.

![image](https://github.com/user-attachments/assets/c18538f4-0d7a-43fb-8579-f200c5ff4381)

I believe this is the real bug here, and perhaps that needs it's own issue, if it's not there already.

**Comment by Kellen275** [comment]
I share the same sentiment as @johanot describes above. I discovered this issue because I frequently have users soft-locking themselves and needing admins to step in to fix their namespace:
1. They run some deployment script
2. It errors saying they can't deploy to their own namespace
3. They ask me why they can't access their own namespace
4. I find that they didn't create their namespace in their project using Rancher. It's orphaned in the "Not in a Project" section.
5. I have to move their namespace into their project

If the initial namespace creation was rejected, we could avoid users getting soft-locked.
